### DISCUSSION TITLE
Is there a way to combine CSR and SSR/SSG for the same app?

### DISCUSSION THREAD

**Discussion by Suedo** [vercel/next.js#12346] [Help] [answered]
My Team is looking to build something like a [Medium](https://medium.com/) clone. 

Here, The posts created by users are not expected to change much, so SSG seems like a good option, especially with the new Incremental build feature (#11552) . However, things like 'upvotes' on posts can change very frequently, so we are unsure how to accommodate that with SSG/SSR.

Also, a portion of the app will lie beyond a login, like Dashboards, or an editor to create a new post... etc. Since these don't need SEO, so we were thinking if that could be done via CSR, while other parts remained SSG or SSR. 

Can such a combination be done with NextJS ? We have been making CSR apps only till now, so apologies if the questions are a bit dumb :( 

Some advice/ideas would be of great help :)

**Comment by SoorajChandran** [comment, marked as answer]
Yes, you can. 
If you use `getServerSideProps` on a page it will be server-rendered. 
Then can use `useSWR` for client-side rendering.

More reading here:
https://nextjs.org/docs/basic-features/data-fetching#fetching-data-on-the-client-side

**Reply by Suedo** [reply]
Awesome! This leads me to ask:

Can i have both CSR and SSG/SSR functionality on the SAME page ? using a combination of `getServerSideProps/getStaticProps` and `useSWR`? 

Like, the content (which needs SEO) is fetched by `getServerSideProps` or `getStaticProps`.. and then, on the client browser, `useSWR` builds on top of it by rendering the upvotes count, via CSR ?

**Reply by SoorajChandran** [reply]
Yes, that is the idea of Next js as far as I understand.

**Reply by Ismoilov** [reply]
> Awesome! This leads me to ask:
> 
> Can i have both CSR and SSG/SSR functionality on the SAME page ? using a combination of `getServerSideProps/getStaticProps` and `useSWR`?
> 
> Like, the content (which needs SEO) is fetched by `getServerSideProps` or `getStaticProps`.. and then, on the client browser, `useSWR` builds on top of it by rendering the upvotes count, via CSR ?

I have same question any answers???

**Reply by Suedo** [reply]
@Ismoilov yes, we can do that. I have tried it and it works

**Reply by Vadorequest** [reply]
@Suedo If you want to take a look at code example and demo, check this:

It uses both SSG/SSR with a per-page approach. 

You can't combine SSG/SSR on the same page, but you can indeed combine server side pre-fetch and then client side fetch.

Demo: https://nrn-v2-mst-aptd-gcms-lcz-sty-c1.now.sh

Implementation: https://github.com/UnlyEd/next-right-now/tree/v2-mst-aptd-gcms-lcz-sty/src/pages/%5Blocale%5D/examples/native-features

**Reply by Nuna7** [reply]
> Yes, you can. If you use `getServerSideProps` on a page it will be server-rendered. Then can use `useSWR` for client-side rendering.
> 
> More reading here: https://nextjs.org/docs/basic-features/data-fetching#fetching-data-on-the-client-side

But how can we do in Angular?

**Comment by chaddjohnson** [comment]
Is it possible to have pure CSR (not SSR) with Next.js?

For example, the route /products/fabf78c5a945b1a67cf11ac8a07067cd would be accessible dynamically via an error page fallback to index.html in CDN settings.

This works great with Create React App, and I would love to see this in Next.js!

Currently an export with Next.js creates one .html file per page, and I believe pure CSR would require just one top-level index.html file providing client-side routing.

**Reply by martinbavio** [reply]
@chaddjohnson you can have a single `[[...catchRootAndAll]].js` file in the `pages` folder and let that be your only entry point to the app.

**Reply by chaddjohnson** [reply]
(Sorry for duplicate threads. I meant to reply to your comment specifically. I've deleted the other)

Thank you @marbiano. For me this results in the following:

<img width="608" alt="Screen Shot 2020-08-13 at 9 51 25 PM" src="https://user-images.githubusercontent.com/676134/90203640-3bc4f400-ddaf-11ea-84e6-9df9efd34ab1.png">
<img width="696" alt="Screen Shot 2020-08-13 at 9 51 43 PM" src="https://user-images.githubusercontent.com/676134/90203641-3c5d8a80-ddaf-11ea-9f7e-d71881edd5ee.png">

No root page loads, and no other pages load. Also, /products/test-product results in a 404.

I have a test repository [here](https://github.com/chaddjohnson/next-ssg-csr/tree/catch-all). Thanks for your time!

**Reply by martinbavio** [reply]
@chaddjohnson You think you can update the repo with your new code. I'm seeing a bunch of other pages in there right now.

**Reply by chaddjohnson** [reply]
Whoops, my apologies @marbiano; I'm not sure how that happened. It's [updated](https://github.com/chaddjohnson/next-ssg-csr/tree/catch-all) now.

**Reply by martinbavio** [reply]
It's working just fine to me, pal.

![Screen Shot 2020-08-14 at 12 20 28 AM](https://user-images.githubusercontent.com/5664/90210318-08409480-ddc4-11ea-9f74-0ec3be6fd78f.png)

![Screen Shot 2020-08-14 at 12 20 39 AM](https://user-images.githubusercontent.com/5664/90210322-0a0a5800-ddc4-11ea-8ff7-b61cca624541.png)

**Reply by chaddjohnson** [reply]
Thanks for checking. I had also run build and export (`yarn export`) and then served the result (`yarn serve`). This is where the problem lies — I need to statically export and then host purely on a CDN with no server.

**Reply by martinbavio** [reply]
I think you need to use `yarn build` and then `yarn export` to export statically. Might want to take a look a Netlify's articles for static Next content: https://www.netlify.com/blog/2020/05/04/building-a-markdown-blog-with-next-9.4-and-netlify/

**Reply by chaddjohnson** [reply]
The `export` script does both: `NODE_ENV=production next build && next export`.

The problem with the approach outlined in the Netlify article is that it expects there to be a complete list of all pages available at build time at build time. So if I use Next for an admin type site and I add a new product at runtime, accessible via /products/:id (and the ID is created when I create the new product), I cannot access the new product I added when the site is hosted purely on a CDN with no server backend (no SSR, no use of getServerSideProps, only using getStaticPaths and getStaticProps).

Using Create React App, the entire site uses only one .html file, /index.html, and I can configure the CDN to use /index.html as an error page fall back. Under this setup, typing http://mydomain.com/products/123 will cause the CDN to load the top-level index.html file, and then client-side routing will kick in (and the URL will remain the same as what I entered) and load the product page.

This is not currently feasible with Next.js as Next exports individual .html files for each page. In order for the setup I described to work, there would need to be a way to optionally export just one top-level index.html file. Maybe via a `target: CDN` configuration or something.

**Comment by vogelino** [comment]
Is there a way to use SSR only when loading a page initially (full page reload) and to fall back to useSWR when switching between pages? 

My point is that as I navigate between pages, the page a click originates from is frozen as long as the `getServerSideProps` function is executed on the client side. In this laps of time, users have to wait without feedback for the page to be displayed. 

I would like to instantly show the page layout with loading skeletons to the users while the `getServerSideProps` function is loading and then pass the results of the `getServerSideProps` to my components (But only while transitioning between pages on the client side). 

I cannot use `getStaticProps` in my case as I'm using csrf tokens and I need access to the request and response objects for each request. Unless there is a was to load my initial data in `getStaticProps` and then my csrf token in `getServerSideProps` but as far as I understand it's either one not both. Right?

**Comment by tbmfg** [comment]
This is what I've done.
```
const fetchData = async () => {
    const res = await fetch("https://jsonplaceholder.typicode.com/todos/1");
    const data = await res.json();
    return data || {};
};

export default function Home({ isCrawler, data: serverData }) {
  const [clientData, setClientData] = useState({});

  useEffect(() => {
    if (!isCrawler) {
      fetchData().then((data) => setClientData(data));
    }
  }, [isCrawler]);

  const renderData = isCrawler ? serverData : clientData;

  return (
    <div className={styles.container}>
      <Head>
        <title>Create Next App</title>
        <meta name="description" content="Generated by create next app" />
        <link rel="icon" href="/favicon.ico" />
      </Head>
      <div>
        <h1>Is Crawler: {isCrawler ? "Yes" : "No"}</h1>
        <h2>Data: {renderData?.title}</h2>
      </div>
    </div>
  );
}

const crawlerAgentRegex =
  /Googlebot|bingbot|Slurp|DuckDuckBot|Baiduspider|YandexBot|Sogou|Exabot|facebot|ia_archiver|Twitterbot|developers\.google\.com\/\+\/web\/snippet\/|Google-Structured-Data-Testing-Tool|LinkedInBot|embedly|quora link preview|showyoubot|outbrain|pinterest\/0\.|pinterestbot|Slack-ImgProxy|Slackbot|vkShare|W3C_Validator|whatsapp/;

export async function getServerSideProps(context) {
  const isCrawler = crawlerAgentRegex.test(context.req.headers["user-agent"]);

  if (isCrawler) {
    const data = await fetchData();

    return {
      props: {
        data,
      },
    };
  }

  return {
    props: {
      data: null,
    },
  };
}
```
